x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/shards/ShardsCollector.java

```diff
@@ -46,24 +45,27 @@ protected Collection<MonitoringDoc> doCollect(final MonitoringDoc.Node node, fin
         if (clusterState != null) {
             RoutingTable routingTable = clusterState.routingTable();
             if (routingTable != null) {
-                List<ShardRouting> shards = routingTable.allShards();
-                if (shards != null) {
-                    final String clusterUuid = clusterUuid(clusterState);
-                    final String stateUUID = clusterState.stateUUID();
-                    final long timestamp = timestamp();
+                final String clusterUuid = clusterUuid(clusterState);
+                final String stateUUID = clusterState.stateUUID();
+                final long timestamp = timestamp();
 
-                    final String[] indices = getCollectionIndices();
-                    final boolean isAllIndices = IndexNameExpressionResolver.isAllIndices(Arrays.asList(indices));
+                final String[] indicesToMonitor = getCollectionIndices();
+                final boolean isAllIndices = IndexNameExpressionResolver.isAllIndices(Arrays.asList(indicesToMonitor));
+                final String[] indices = isAllIndices ? routingTable.indicesRouting().keySet().toArray(new String[0]) : indicesToMonitor;
 
-                    for (ShardRouting shard : shards) {
-                        if (isAllIndices || Regex.simpleMatch(indices, shard.getIndexName())) {
-                            MonitoringDoc.Node shardNode = null;
-                            if (shard.assignedToNode()) {
-                                // If the shard is assigned to a node, the shard monitoring document refers to this node
-                                shardNode = convertNode(node.getTimestamp(), clusterState.getNodes().get(shard.currentNodeId()));
-                            }
-                            results.add(new ShardMonitoringDoc(clusterUuid, timestamp, interval, shardNode, shard, stateUUID));
+                for (String index : indices) {
+                    int shardCount = 0;
+                    for (ShardRouting shard : routingTable.allShards(index)) {
```

Comment:
One issue I faced while working on the feedback, is that the list of indices to monitor can include pattern blobs as well.

Is there a idiomatic way in the ES code base to expand an index pattern (like `metricbeat-*`) to a list of indices that match that pattern based on the current routing table? I wrote some hand rolled code for doing this but I suspect this is a more common operation.

Reply:
They were supported since we looped over every index and used the `Regex.simpleMatch` to see if a pattern matched the index. In the code I wrote to expand patterns I use the same method.